numba/core/dispatcher.py

```diff
@@ -687,7 +687,10 @@ def typeof_pyval(self, val):
         try:
             tp = typeof(val, Purpose.argument)
         except ValueError:
-            tp = types.pyobject
+            if hasattr(val, '__array__'):
+                tp = typeof(val.__array__(), Purpose.argument)
```

Comment:
I think the CI failure calls to attention that I need an additional `try: ... except ValueError: ...` around this, but that wouldn't be sufficient to make this work, due to #9585 - the `NotImplementedError` it produces would not be caught by `except ValueError`.